src/syntax/syntax.rkt

```diff
@@ -83,32 +112,130 @@
+         (unless (equal? ift-ty iff-ty)
+           (type-error! iff iff-ty ift-ty))
+         ift-ty]
+        [`(,op ,args ...)
+         (unless (operator-exists? op)
+           (bad! "expected operator at `~a`, got `~a` in `~a`" expr op))
+         (define itypes (operator-info op 'itype))
+         (for ([arg (in-list args)] [itype (in-list itypes)])
+           (define arg-ty (type-of arg))
+           (unless (equal? itype arg-ty)
+             (type-error! arg arg-ty itype)))
+         (operator-info op 'otype)]
+        [_ (bad! "expected an expression, got `~a`" expr)])))
+
+  (unless (equal? actual-ty otype)
+    (type-error! body actual-ty otype)))
+
+;; Applies a substitution.
+;; Slightly different than `replace-vars` in `programs.rkt`.
+(define (replace-vars expr env)
+  (let loop ([expr expr])
+    (match expr
+      [(? number?) expr]
+      [(? symbol?) (cdr (assq expr env))]
+      [`(,op ,args ...) `(,op ,@(map loop args))])))
+
+;; Expands an "accelerator" specification.
+;; Any nested accelerator is unfolded into its definition.
+(define (expand-accelerator-spec spec)
+  (let loop ([expr spec])
+    (match expr
+      [(? number?) expr]
+      [(? symbol?) expr]
+      [`(,(? operator-accelerator? op) ,args ...)
+       (define spec (operator-info op 'spec))
+       (match-define `(,(or 'lambda 'λ) (,vars ...) ,body) spec)
+       (define env (map cons vars (map loop args)))
+       (replace-vars body env)]
+      [`(,op ,args ...) `(,op ,@(map loop args))])))
+
 ;; Registers an operator with an attribute mapping.
 ;; Panics if an operator with name `name` has already been registered.
-;; By default, the input types are specified by `itypes`,
-;; the output type is specified by `otype`, and the operator is not
-;; deprected, but `attrib-dict` can override these properties.
+;; By default, the input types are specified by `itypes`, the output type
+;; is specified by `otype`, and the operator is not deprecated; but
+;; `attrib-dict` can override these properties.
 (define (register-operator! name itypes otype attrib-dict)
   (when (hash-has-key? operators name)
     (error 'register-operator! "operator already registered: ~a" name))
-  (define attribs
-    (hash 'itype
-          (dict-ref attrib-dict 'itype itypes)
-          'otype
-          (dict-ref attrib-dict 'otype otype)
-          'deprecated
-          (dict-ref attrib-dict 'deprected #f)
-          'ival
-          (dict-ref attrib-dict
-                    'ival
-                    (λ () (error 'register-operator! "missing interval impl for ~a" name)))))
-  (define field-names '(itype otype ival deprecated))
-  (define table-entry (apply operator name (map (curry hash-ref attribs) field-names)))
-  (hash-set! operators name table-entry)
+  ; extract relevant fields
+  (define itypes* (dict-ref attrib-dict 'itype itypes))
+  (define otype* (dict-ref attrib-dict 'otype otype))
+  (define ival-fn (dict-ref attrib-dict 'ival #f))
```

Comment:
Again, we don’t actually need this

Reply:
Oh right. Forgot that we don't use `ival` anymore. I initially created `src/core/rival.rkt` to get something minimal to include in `src/syntax/syntax.rkt` but I guess that's no longer the case.